src/solana/solanaEngine.js

```diff
@@ -0,0 +1,390 @@
+    const fee = tx.meta.fee
+    if (amount < 0) {
+      amount -= fee
+    } else {
+      ourReceiveAddresses.push(this.keypair.publicKey.toBase58())
+    }
+    const edgeTransaction: EdgeTransaction = {
+      txid: tx.transaction.signatures[0],
+      date: timestamp,
+      currencyCode: this.chainCode,
+      blockHeight: tx.slot,
+      nativeAmount: amount.toString(),
+      networkFee: fee.toString(),
+      ourReceiveAddresses,
+      signedTx: ''
+    }
+    this.addTransaction(this.chainCode, edgeTransaction)
+  }
+
+  async queryTransactions() {
+    let before = null
+    const until =
+      this.otherData.newestTxid !== '' ? this.otherData.newestTxid : null
+    let txids = []
+    try {
+      // Gather all transaction IDs since we last updated
+      while (1) {
+        const params = [
+          this.keypair.publicKey.toBase58(),
+          {
+            until,
+            before,
+            limit: this.settings.txQueryLimit,
+            commitment: this.settings.commitment
+          }
+        ]
+        const response: RpcSignatureForAddress[] = await this.fetchRpc(
+          'getSignaturesForAddress',
+          params
+        )
+
+        // remove failed transactions from list
+        txids = txids.concat(
+          response.filter(data => data.err == null && data.signature != null)
+        )
+
+        if (response.length < this.settings.txQueryLimit) break // RPC limit
+        before = response[this.settings.txQueryLimit - 1].signature
+      }
+    } catch (e) {
+      this.error('getTransactionSignatures failed with error: ', e)
+      return
+    }
+
+    if (txids.length === 0) {
+      this.tokenCheckTransactionsStatus[this.chainCode] = 1
+      this.updateOnAddressesChecked()
+      return
+    }
+
+    let newestTxid = 0
+    for (let i = 0; i < txids.length; i++) {
+      try {
+        const tx = asRpcGetTransaction(
+          await this.fetchRpc('getTransaction', [
+            txids[i].signature,
+            { encoding: 'json', commitment: this.settings.commitment }
+          ])
+        )
+
+        // From testing, getSignaturesForAddress always returns a blocktime but it is optional in the RPC docs so we should be prepared to get it if it isn't present
+        let timestamp = txids[i].blocktime
+        if (timestamp == null || isNaN(timestamp))
+          timestamp = asNumber(await this.fetchRpc('getBlockTime', [tx.slot]))
+
+        this.processSolanaTransaction(tx, timestamp)
+      } catch (e) {
+        // Note the oldest failed tx query so we try again next loop
+        newestTxid = i
+      }
+    }
+
+    // Don't update the newestTxid if the txids array length is length 1 or the oldest failed query is the end of the array.
+    // The previously saved value is the best to use in those cases
+    if (txids.length > 1 && txids.length > newestTxid + 1)
+      this.otherData.newestTxid = txids[newestTxid + 1].signature
```

Comment:
`newestTxid` has an inconsistent type. sometimes a number, sometimes a string. make it consistent.